### PR TITLE
[PyGrain Migration] Run RandomDeletion and RandomSwap in Grain pipelines

### LINKED ISSUE #2952: [PyGrain Migration] 8. Migrate RandomDeletion / RandomSwap
- Implement Python paths operating on lists-of-token-lists (mirror the tf logic, including skip_list / skip_fn / skip_py_fn options).
- Seeding: derive an np.random.Generator per record from (self.seed, stable_hash(record)) so results are deterministic regardless of worker count, and consistent between direct calls and Grain runs.
- Add determinism tests: same seed + same record -> same output across two separate pipeline runs and across num_workers=1 vs num_workers=2.

### PR DESCRIPTION
**Pull request by laxmareddyp**
This Fixes #2952 

Add a pure Python/NumPy execution path to both layers so they can run in Grain worker processes without TensorFlow, following the existing `_call_python` / `_call_tf` dispatch pattern. `_allow_python_workflow` now defaults to `True`.

Both layers previously built a `tf.random.Generator` in `__init__`. Grain pickles the layer into every worker, so that generator is copied and each worker would replay the same stream. Randomness is now derived per record from the layer seed plus a stable content hash, so output is independent of worker count, worker assignment and record position. The hash uses blake2b rather than the builtin `hash()`, which is salted per process by `PYTHONHASHSEED`.

Content derived seeding augments an identical record identically on every epoch, so `call` also accepts an optional `rng` generator. Users can wrap the layer in `grain.RandomMapTransform` to get per epoch varying and worker independent randomness. Passing `rng` on the TensorFlow path raises instead of silently ignoring it.

`tf.random.Generator` and `tf.lookup.StaticHashTable` are now built lazily by the TensorFlow path only, so construction no longer requires TensorFlow. `skip_list` is also kept as a Python set for the NumPy path. `skip_fn` requires traceable TF ops, so it keeps dispatching to the TensorFlow path.

The NumPy path cannot reproduce TensorFlow's RNG stream, so the RNG dependent docstring and test expectations are regenerated. Seeds were changed where the previous seed produced a vacuous example. Tests also assert invariants now rather than only exact strings.

## Checklist

- [x] I have read and followed the [PR contribution policy](https://github.com/keras-team/keras/issues/23601).
- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [x] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

File: keras_hub/src/utils/random_utils_test.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    def test_hash_is_stable_across_processes(self):
+        # `PYTHONHASHSEED` is randomized per process by default, which is
+        # exactly what would break Grain worker determinism if we used the
+        # builtin `hash()`.
+        module_path = random_utils_module.__file__
+        script = (
+            "import importlib.util;"
+            f"spec = importlib.util.spec_from_file_location('ru', "
+            f"{module_path!r});"
```
Comment (divyashreepathihalli) [inline keras_hub/src/utils/random_utils_test.py:33]: Hashing non-string tokens by repr makes the seed depend on the Python container type, not the content. On this head, RandomDeletion(rate=0.5, seed=7) gives [5] for [1, ..., 8] and for np.array([1, ..., 8]), but [3, 4, 7, 8] for the same values as a list of np.int64, because repr(np.int64(1)) is 'np.int64(1)' under NumPy 2 and '1' under NumPy 1. That breaks the "same record, same augmentation" guarantee across input types and across NumPy versions. Normalize numbers before hashing, e.g. token.item() for np.generic, then hash int/float canonically (str(int(token)), or struct.pack for floats).

File: keras_hub/src/layers/preprocessing/random_deletion_test.py [hunk 2]
```diff
@@ -141,3 +147,77 @@ def skip_py_fn(word):
...
+        # matter when, where, or in what batch it is seen.
+        augmenter = RandomDeletion(rate=0.5, seed=42)
+        data = [[f"tok{i}", "a", "b", "c", "d"] for i in range(8)]
+        forward = [augmenter(row) for row in data]
+        backward = [augmenter(row) for row in reversed(data)]
+        self.assertEqual(forward, list(reversed(backward)))
+        # Batched and unbatched calls agree too.
+        self.assertEqual(augmenter(data), forward)
+
+    @pytest.mark.large
```
Comment (divyashreepathihalli) [inline keras_hub/src/layers/preprocessing/random_deletion_test.py:174]: did you run and test this locally? not sure if this large test is running on the cI

File: keras_hub/src/layers/preprocessing/random_deletion.py [hunk 3]
```diff
@@ -40,40 +52,48 @@ class RandomDeletion(PreprocessingLayer):
...
+    Call arguments:
+        inputs: The tokens to augment.
+        rng: Optional `np.random.Generator` used instead of the per record
+            generator derived from `seed`. Only supported on the pure Python
+            code path.
+
     Examples:
 
     Word level usage.
     >>> keras.utils.set_random_seed(1337)
```
Comment (divyashreepathihalli) [inline keras_hub/src/layers/preprocessing/random_deletion.py]: These calls no longer affect the output on the default path, since randomness is derived from seed plus record content. Leaving them in the examples suggests the global seed matters. Remove them from the examples, keep them only in the skip_fn example that still runs on the TF path.

File: keras_hub/src/layers/preprocessing/random_deletion.py [hunk 4]
```diff
@@ -239,6 +280,68 @@ def _shuffle_and_trim(x):
...
+                [token for i, token in enumerate(row) if i not in deleted]
+            )
+
+        if not batched:
+            outputs = outputs[0]
+        # Deletion always produces ragged output, which the TensorFlow path
+        # also returns as nested python lists.
+        return outputs
+
+    def call(self, inputs, rng=None):
```
Comment (divyashreepathihalli) [inline keras_hub/src/layers/preprocessing/random_deletion.py:333]: rng is new public API on call. It works well with grain.MapDataset.random_map , so please add that exact snippet to the class docstring, since it is the recommended way to get per-epoch variation. Also worth a line in the docstring that with a batched input all rows draw from the one passed generator in order.

File: keras_hub/src/layers/preprocessing/random_deletion.py [hunk 5]
```diff
@@ -239,6 +280,68 @@ def _shuffle_and_trim(x):
...
+        # `skip_fn` is documented to consist of TensorFlow ops, so a layer
+        # configured with it can only run on the TensorFlow path.
+        use_tf = (
+            not self._allow_python_workflow
+            or self.skip_fn is not None
+            or in_tf_function()
+        )
```
[the remark this answers is not in the document]
Comment (divyashreepathihalli) [reply keras_hub/src/layers/preprocessing/random_deletion.py:340]: Not valid. Both layers define compute_output_shape, so Keras never calls call with symbolic inputs.

File: keras_hub/src/layers/preprocessing/random_swap.py [hunk 6]
```diff
@@ -235,6 +276,71 @@ def _swap(x):
...
+        return outputs
+
+    def call(self, inputs, rng=None):
+        # `skip_fn` is documented to consist of TensorFlow ops, so a layer
+        # configured with it can only run on the TensorFlow path.
+        use_tf = (
+            not self._allow_python_workflow
+            or self.skip_fn is not None
+            or in_tf_function()
+        )
```
[the remark this answers is not in the document]
Comment (divyashreepathihalli) [reply keras_hub/src/layers/preprocessing/random_swap.py:339]: Not valid. Both layers define compute_output_shape, so Keras never calls call with symbolic inputs.

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]

File: keras_hub/src/layers/preprocessing/random_deletion_test.py (hunk 2 above)
Comment (laxmareddyp) [reply keras_hub/src/layers/preprocessing/random_deletion_test.py:174]: It is already ran in GPU test suite.

File: keras_hub/src/utils/random_utils_test.py (hunk 1 above)
Comment (laxmareddyp) [reply keras_hub/src/utils/random_utils_test.py:33]: Fixed. Numbers now hash by value regardless of container type. I also fixed a regression where negative seeds would crash `SeedSequence`.

File: keras_hub/src/layers/preprocessing/random_deletion_test.py (hunk 2 above)
Comment (laxmareddyp) [reply keras_hub/src/layers/preprocessing/random_deletion_test.py:174]: It runs in the Kokoro GPU build, which explicitly passes `--run_large`. It isn't covered by the GitHub Actions CPU matrix.

File: keras_hub/src/layers/preprocessing/random_deletion.py (hunk 3 above)
Comment (laxmareddyp) [reply keras_hub/src/layers/preprocessing/random_deletion.py]: Removed `set_random_seed` from all non-TF examples across both `random_deletion.py` and `random_swap.py`.

File: keras_hub/src/layers/preprocessing/random_deletion.py (hunk 4 above)
Comment (laxmareddyp) [reply keras_hub/src/layers/preprocessing/random_deletion.py:333]: Added a runnable doctest showing how to use `rng` with `grain.MapDataset.random_map` in both layers.

**Review by divyashreepathihalli — APPROVED** [review]

File: keras_hub/src/layers/preprocessing/random_swap.py [hunk 7]
```diff
@@ -144,22 +183,33 @@ def __init__(
                 "provided."
             )
 
-        if self.skip_list:
-            self.StaticHashTable = tf.lookup.StaticHashTable(
+        self._skip_set = set(self.skip_list) if self.skip_list else None
```
[the remark this answers is not in the document]
Comment (divyashreepathihalli) [reply keras_hub/src/layers/preprocessing/random_swap.py]: +1

File: keras_hub/src/layers/preprocessing/random_deletion.py [hunk 8]
```diff
@@ -148,22 +187,33 @@ def __init__(
                 "provided."
             )
 
-        if self.skip_list:
-            self.StaticHashTable = tf.lookup.StaticHashTable(
+        self._skip_set = set(self.skip_list) if self.skip_list else None
```
[the remark this answers is not in the document]
Comment (divyashreepathihalli) [reply keras_hub/src/layers/preprocessing/random_deletion.py]: +1

**Comment by divyashreepathihalli** [comment]
Please address the Gemini comments before merge! Thanks!
